### ISSUE TITLE
IParameterReferenceExpression operation not generated for parameter access in a certain cases

### ISSUE THREAD

**Issue by mavasani** [dotnet/roslyn#8884]
Sample test code:

``` C#
using System;

public class NeatCode
{
    // Used parameter methods
    public void UsedParameterMethod1(Action a)
    {
        Action a2 = new Action(() =>
        {
            a(); // No parameter reference expression operation generated for this access
        });
    }
}
```

This does not repro for VB.

**Comment by mavasani** [comment]
**Quick repro:** Build roslyn-analyzers into RoslynDev hive and create a project with above code - you should see CA1801 (remove unused parameter) warning.

**Comment by mavasani** [comment]
Found couple more cases where this repros:

1. Parameter access in a using statement (repros for both VB and C#)
2. Parameter access in a linq statement (repros for both VB and C#)

```C#
using System;
using System.Reflection;

class C
{
    void F(int x, IDisposable o)
    {
        using (o)
        {
            int y = x; // no parameter reference received for 'x'
        }
    }

    private object F2(Assembly assembly)
    {
        return (from t in assembly.GetTypes()
                    select t.Attributes).FirstOrDefault();  // no parameter reference received for 'assembly'
    }
}
```

**Comment by mavasani** [comment]
This is causing lot of false reports in Analyzers.sln in roslyn-analyzers repo - we should probably test the fix for this bug against that solution.

**Comment by mavasani** [comment]
We get no parameter reference for parameter type if used in a string as follows: `$"{type.Name}"`

**Comment by bkoelman** [comment]
Yes, I also get false reports for CA1801 in interpolated strings and LINQ queries. 

Additionally, I see unexpected reports on parameter usage in initializers:

```csharp
        private Panel CreateRankingsOverlay(string text)
        {
            return new Panel
            {
                Controls =
                {
                    new AutoScalingLabel
                    {
                        Text = text,
                    }
                }
            };
        }
```

```csharp
            private static char AutoDetectFieldSeparator(string line)
            {
                var charCounts = new Dictionary<char, int>
                {
                    { '\t', line.Count(c => c == '\t') },
                    { ';', line.Count(c => c == ';') },
                    { ',', line.Count(c => c == ',') },
                    { ':', line.Count(c => c == ':') },
                    { '|', line.Count(c => c == '|') }
                };

                int highestOccurrence = charCounts.Max(pair => pair.Value);
                return charCounts.First(pair => pair.Value == highestOccurrence).Key;
            }
```

**Comment by mavasani** [comment]
@cston found another case for pointer expressions:

```
unsafe class C
{
    static int F(byte* b, int x, int y)
    {
        return b[x] + y;
    }
}
```

No parameter reference generated for `b` and `x`.

**Comment by mavasani** [comment]
Re-activating as we still have couple of cases which have not been addressed with #18976. More specifically, parameter references in delegate creation and collection initializers.

**Comment by AlekseyTs** [comment]
It looks like the problem still exists in some scenarios:
```

        <CompilerTrait(CompilerFeature.IOperation)>
        <Fact, WorkItem(8884, "https://github.com/dotnet/roslyn/issues/8884")>
        Public Sub ParameterReference_NoPiaObjectCreation()
            Dim sources0 = <compilation>
                               <file name="a.vb"><![CDATA[
Imports System.Runtime.InteropServices
<Assembly: ImportedFromTypeLib("_.dll")>
<Assembly: Guid("f9c2d51d-4f44-45f0-9eda-c9d599b58257")>
<ComImport()>
<Guid("f9c2d51d-4f44-45f0-9eda-c9d599b58277")>
<CoClass(GetType(C))>
Public Interface I
    Property P As Integer
End Interface
<Guid("f9c2d51d-4f44-45f0-9eda-c9d599b58278")>
Public Class C
    Public Sub New(o As Object)
    End Sub
End Class
]]></file>
                           </compilation>
            Dim sources1 = <compilation>
                               <file name="a.vb"><![CDATA[
Structure S
    Function F(x as Object) As I
        Return New I(x)'BIND:"New I(x)"
    End Function
End Structure
]]></file>
                           </compilation>
            Dim compilation0 = CreateCompilationWithMscorlib40(sources0)
            compilation0.AssertTheseDiagnostics()

            ' No errors for /r:_.dll
            Dim compilation1 = CreateEmptyCompilationWithReferences(
                sources1,
                references:={MscorlibRef, SystemRef, compilation0.EmitToImageReference(embedInteropTypes:=True)})

            Dim expectedOperationTree = <![CDATA[
INoPiaObjectCreationOperation (OperationKind.None, Type: I, IsInvalid) (Syntax: 'New I(x)')
  Initializer: 
    null
]]>.Value

            Dim expectedDiagnostics = <![CDATA[
BC30516: Overload resolution failed because no accessible 'New' accepts this number of arguments.
        Return New I(x)'BIND:"New I(x)"
               ~~~~~~~~
]]>.Value

            VerifyOperationTreeAndDiagnosticsForTest(Of ObjectCreationExpressionSyntax)(compilation1, "a.vb", expectedOperationTree, expectedDiagnostics)
        End Sub
```

**Comment by AlekseyTs** [comment]
And for C# as well:
```

        [CompilerTrait(CompilerFeature.IOperation)]
        [Fact, WorkItem(8884, "https://github.com/dotnet/roslyn/issues/8884")]
        public void ParameterReference_NoPiaObjectCreation()
        {
            var sources0 = @"
using System;
using System.Runtime.InteropServices;

[assembly: ImportedFromTypeLib(""_.dll"")]
[assembly: Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58257"")]
[ComImport()]
[Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58277"")]
[CoClass(typeof(C))]
public interface I
        {
            int P { get; set; }
        }
[Guid(""f9c2d51d-4f44-45f0-9eda-c9d599b58278"")]
public class C
{
    public C(object o)
    {
    }
}
";
            var sources1 = @"
struct S
{
	public I F(object x)
	{
		return /*<bind>*/new I(x)/*</bind>*/;
    }
}
";
            var compilation0 = CreateCompilation(sources0);
            compilation0.VerifyDiagnostics();

            var compilation1 = CreateEmptyCompilation(
                sources1,
                references: new[] { MscorlibRef, SystemRef, compilation0.EmitToImageReference(embedInteropTypes: true) });

            string expectedOperationTree = @"
INoPiaObjectCreationOperation (OperationKind.None, Type: I, IsInvalid) (Syntax: 'new I(x)')
  Initializer: 
    null
";
            var expectedDiagnostics = new DiagnosticDescription[] {
                    // (6,25): error CS1729: 'I' does not contain a constructor that takes 1 arguments
                    // 		return /*<bind>*/new I(x)/*</bind>*/;
                    Diagnostic(ErrorCode.ERR_BadCtorArgCount, "(x)").WithArguments("I", "1").WithLocation(6, 25)
            };

            VerifyOperationTreeAndDiagnosticsForTest<ObjectCreationExpressionSyntax>(compilation1, expectedOperationTree, expectedDiagnostics);
        }
```
